### PR TITLE
[DMS-175] Start PostgreSQL database in E2E test framework

### PR DESCRIPTION
**Pull request by JBrenesSimpat**

### CODE REVIEW

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Hooks/SetupHooks.cs [hunk 1]
```diff
@@ -39,6 +39,13 @@ TestLogger logger
                 logger.log.Debug("Using local environment, verify that it's correctly set.");
             }
 
+            string? dbEngine = _configuration["DatabaseEngine"];
+
+            if (dbEngine != null)
+            {
+                context.Database = await containers.SetupDataBase(dbEngine);
+            }
+
```
Comment (stephenfuqua) [inline src/services/EdFi.DataManagementService.Api.Tests.E2E/Hooks/SetupHooks.cs:48]: I see `if (useTestContainers)` above on line 32.

Would it make sense to put your new code into that same branching statement, instead of being below the `else`?

**Comment by simpat-adam** [comment]
I think you might have to modify the connection string in the appsettings.json file from the web api project to use Postgresql to get the E2E tests to pass.

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs [hunk 2]
```diff
@@ -28,4 +28,32 @@ public async Task<string> SetupDataManagement()
...
+        if (engine == "postgresql")
+        {
+            imageName = "local/edfi-data-management-postgresql";
+            port = 5432;
+        }
+
+        var dockerImage = new ContainerBuilder()
+            .WithImage(imageName)
+            .WithPortBinding(port, port)
+            .WithWaitStrategy(Wait.ForUnixContainer().UntilPortIsAvailable(port))
```
Comment (simpat-adam) [inline src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs:46]: When I run locally I have to remove this line, otherwise the tests never start. Maybe I have something misconfigured?

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Features/ConectToPostgresql.feature
Comment (simpat-adam) [inline src/services/EdFi.DataManagementService.Api.Tests.E2E/Features/ConectToPostgresql.feature:1]: I'm not sure this feature test is worth the effort. My branch will sufficiently test all this and more.

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs [hunk 3]
```diff
@@ -28,4 +28,32 @@ public async Task<string> SetupDataManagement()
             dockerImage.GetMappedPublicPort(8080)
         ).ToString();
     }
+
+    public async Task<string> SetupDataBase(string engine)
```
Comment (simpat-adam) [inline src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs:32]: You don't need to return anything from this method. Also I think I would suggest removing all the logic / configuration around `DatabaseEngine` Just assume it's all Postgres for now. Probably rename this method to `SetupPostgresBackend` and keep it simple.

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/PlaywrightContext.cs [hunk 4]
```diff
@@ -13,6 +13,8 @@ public class PlaywrightContext
 
     public string ApiUrl { get; set; } = "http://localhost:5198";
 
+    public string Database { get; set; } = "";
```
Comment (simpat-adam) [inline src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/PlaywrightContext.cs:16]: I don't think you need this.

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/appsettings.json [hunk 5]
```diff
@@ -1,5 +1,6 @@
 {
     "UseTestContainers": true,
+    "DatabaseEngine": "postgresql",
```
Comment (simpat-adam) [inline src/services/EdFi.DataManagementService.Api.Tests.E2E/appsettings.json:3]: See previous comment. I would remove this for now.

**Review by simpat-adam — CHANGES_REQUESTED** [review]

**Comment by simpat-adam** [comment]
It looks to me like the builds are failing because the image `local/edfi-data-management-postgresql` is not being built. I believe you'll have to modify `build.ps1` to achieve that. There is already a .yml file `postgres-deployment.yml`

File: src/frontend/EdFi.DataManagementService.Frontend.AspNetCore/appsettings.json [hunk 6]
```diff
@@ -3,11 +3,11 @@
         "BeginAllowedSchoolYear": 2022,
         "EndAllowedSchoolYear": 2035,
         "AuthenticationService": "http://dev-test-auth-service",
-        "DatabaseEngine": "mssql",
+        "DatabaseEngine": "postgresql",
         "DeployDatabaseOnStartup": false
     },
     "ConnectionStrings": {
-        "DatabaseConnection": "Data Source=.\\;Initial Catalog=master;Integrated Security=True;"
+        "DatabaseConnection": "host=dmsdb;port=5432;username=postgres;database=edfi_datamanagementservice;"
```
Comment (stephenfuqua) [inline src/frontend/EdFi.DataManagementService.Frontend.AspNetCore/appsettings.json:10]: This is probably fine, Generally I wouldn't put Docker-specific connection information into the main `appsettings.json`. We can change later if needed - don't worry about it right now.

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs [hunk 7]
```diff
@@ -11,21 +11,34 @@ public class ContainerSetup
...
+            .WithImage(apiImageName)
             .WithPortBinding(8080)
             .WithWaitStrategy(Wait.ForUnixContainer().UntilHttpRequestIsSucceeded(r => r.ForPort(8080)))
+            .WithNetwork(network)
             .Build();
 
-        await dockerImage.StartAsync();
+        var dbContainer = new ContainerBuilder()
+        .WithImage(dbImageName)
+        .WithPortBinding(5432, 5432)
```
Comment (stephenfuqua) [inline src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs:29]: In another ticket - let's get this one done first - I would like to use a different port number, so that we can run E2E tests locally without conflicting with another instance of PostgreSQL that is already running.

**Review by stephenfuqua — APPROVED** [review]

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs (hunk 7 above)
Comment (stephenfuqua) [reply src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs:29]: Come to think of it, this might not be a problem. I don't know if TestContainers is exposing this port to localhost; perhaps it is only internal on the Docker network.

File: src/backend/EdFi.DataManagementService.Backend.Postgresql/PostgresqlDocumentStoreRepository.cs [hunk 8]
```diff
@@ -49,8 +49,9 @@ public async Task<UpsertResult> UpsertDocument(IUpsertRequest upsertRequest)
             _logger.LogError("Unknown Error - {TraceId}", upsertRequest.TraceId);
             return new UpsertResult.UnknownFailure("Unknown Failure");
         }
-        catch
+        catch (Exception ex)
```
Comment (simpat-adam) [inline src/backend/EdFi.DataManagementService.Backend.Postgresql/PostgresqlDocumentStoreRepository.cs:52]: @JBrenesSimpat Please revert this change I pushed for debugging.

**Review by simpat-adam — CHANGES_REQUESTED** [review]

File: src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs (hunk 7 above)
Comment (simpat-adam) [reply src/services/EdFi.DataManagementService.Api.Tests.E2E/Management/ContainerSetup.cs:29]: @stephenfuqua I haven't tried it since I added that network wrapper around our test containers. Before then though - yeah I was starting and stopping a lot.
